### ISSUE TITLE
Program throws exception for simple program

### ISSUE THREAD

**Issue by AVassilev98** [daniel5151/ts7200#25]
Hi there, very cool project! Seeing some issues when trying to test this with a simple program. Program is as follows:

`int main(void) { return 0; }`

I am compiling this using the TS7200 cross compiler provided by TS here: https://files.embeddedarm.com/ts-arm-sbc/ts-7200-linux/cross-toolchains/crosstool-linux-gcc-4.0.1-glibc-2.3.5.tar.bz2

When running `cargo run --release <binary>` I see:

```
 DEBUG ts7200::sys::ts7200 > Setting PC to 0x0000855c
 DEBUG ts7200::sys::ts7200 > loading section ".interp" into memory from [0x000080f4..0x00008107]
 DEBUG ts7200::sys::ts7200 > loading section ".note.ABI-tag" into memory from [0x00008108..0x00008128]
 DEBUG ts7200::sys::ts7200 > loading section ".hash" into memory from [0x00008128..0x000081dc]
 DEBUG ts7200::sys::ts7200 > loading section ".dynsym" into memory from [0x000081dc..0x0000837c]
 DEBUG ts7200::sys::ts7200 > loading section ".dynstr" into memory from [0x0000837c..0x0000848e]
 DEBUG ts7200::sys::ts7200 > loading section ".gnu.version" into memory from [0x0000848e..0x000084c2]
 DEBUG ts7200::sys::ts7200 > loading section ".gnu.version_r" into memory from [0x000084c4..0x000084e4]
 DEBUG ts7200::sys::ts7200 > loading section ".rel.dyn" into memory from [0x000084e4..0x0000850c]
 DEBUG ts7200::sys::ts7200 > loading section ".rel.plt" into memory from [0x0000850c..0x0000851c]
 DEBUG ts7200::sys::ts7200 > loading section ".init" into memory from [0x0000851c..0x00008530]
 DEBUG ts7200::sys::ts7200 > loading section ".plt" into memory from [0x00008530..0x0000855c]
 DEBUG ts7200::sys::ts7200 > loading section ".text" into memory from [0x0000855c..0x00008780]
 DEBUG ts7200::sys::ts7200 > loading section ".fini" into memory from [0x00008780..0x0000878c]
 DEBUG ts7200::sys::ts7200 > loading section ".rodata" into memory from [0x0000878c..0x00008790]
 DEBUG ts7200::sys::ts7200 > loading section ".eh_frame" into memory from [0x00008790..0x00008794]
 DEBUG ts7200::sys::ts7200 > loading section ".ctors" into memory from [0x00010794..0x0001079c]
 DEBUG ts7200::sys::ts7200 > loading section ".dtors" into memory from [0x0001079c..0x000107a4]
 DEBUG ts7200::sys::ts7200 > loading section ".jcr" into memory from [0x000107a4..0x000107a8]
 DEBUG ts7200::sys::ts7200 > loading section ".dynamic" into memory from [0x000107a8..0x00010870]
 DEBUG ts7200::sys::ts7200 > loading section ".got" into memory from [0x00010870..0x00010898]
 DEBUG ts7200::sys::ts7200 > loading section ".data" into memory from [0x00010898..0x000108a4]
 WARN  ts7200::sys::ts7200 > [pc 0x00008568][addr 0x01fdcf34][Ram > <data>] r32 from uninitialized RAM
 WARN  ts7200::sys::ts7200 > [pc 0x2d2d2d2d][addr 0x00000020][Ram > <data>] r32 from uninitialized RAM
 ERROR ts7200              > Fatal Error! Caused by: FatalMemException {
    addr: 0x2d2d2d2c,
    in_mem_space_of: "<unmapped>",
    reason: Unexpected,
}
 ERROR ts7200              > Dumping system state to sysdump.log
Error: "Fatal Error!"
```

Do you have any ideas as to why this might be happening? Is this a bug, or some mistake I have made in setting up?

**Comment by daniel5151** [comment]
Hey there! Glad to see that CS 452 is back again!
I was really worried it wouldn't survive the pandemic 😅

I take it you're eager to get a bit of a head start on things? I was the same way lol - I wrote v0 of ts7200 on co-op the term before I took CS 452.

* * *

I think the reason things aren't working for you using the public toolchain is that `ts7200` isn't a very "fleshed out" emulator that's designed to run general purpose TS-7200 software. It's very tailored to the specific featureset used by CS 452. This extends to the specific memory configuration that the Redboot bootloader sets up for the software.

As such, when compiling your code, you'll need to use a custom linker script to lay out the sections in memory correctly.

Take a look at the example code on the school servers at `/u/cs452/public/iotest/` (linked to at https://student.cs.uwaterloo.ca/~cs452/W22/), and make sure you are able to build + run that code in ts7200.

**Comment by daniel5151** [comment]
fwiw, here is the memory map of the emulator: https://github.com/daniel5151/ts7200/blob/master/src/sys/ts7200/mod.rs#L401-L412

Bad Things :tm: will happen if you access memory outside those ranges.

**Comment by AVassilev98** [comment]
Aha, yeah pretty much. Just want to get my development environment set up before the course starts so I don't have to worry about it.

Thanks for the advice, looking at the memory map it makes sense why things were failing before. I followed your suggestion and got the CS452 compiler from [here](https://student.cs.uwaterloo.ca/~cs452/W22/COMPILE.md) and linker script from [here](https://student.cs.uwaterloo.ca/~cs452/W22/iotest.xz) (Would it be a good idea to link to these in the README.md?).

I was able to run my simple program on your emulator pretty easily after getting those, so I'm closing this issue. Thanks for the help!
